# Changelog

All notable changes to the `picard_qc` module will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0] - 2020-06-08

This release was authored by Helena Winata.

- The picard metrics calculated for each sample depends on `seq-type`:
    - mrna: alignment summary,insert size and rnaseq metrics.
    - capture: alignment summary,insert size, and hs metrics.
    - genome: alignment summary,insert size and wgs metrics.
- Flagstats are also calculated via samtools for each sample.
- The final outputs are merged metrics from all samples within the same `seq_type` and `genome_build`. 
- Users are required to specify `CFG['switches']['capture_intervals']`, which refers to a text file containing the capture regions used in sequencing.
    - The files in `config["lcr-modules"]["picard_qc"]["switches"]["capture_intervals"]` has to correspond to `config["lcr-modules"]["utils"]["inputs"]["bed"] if intervals are provided in BED format`
    - For demo purposes run:
    ```bash
    touch demo/reference/exomes/grch37/bed/S07604624.bed
    ```
- Reference files for CollectRNaSeqMetrics (rRNA interval list and refFlat files) generated by `reference_files` subworkflow.
