# Changelog

All notable changes to the `picard_qc` module will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0] - 2020-06-08

This release was authored by Helena Winata.

- The picard metrics calculated for each sample depends on `seq-type`:
    - mrna: alignment summary,insert size and rnaseq metrics.
    - capture: alignment summary,insert size, and hs metrics.
    - genome: alignment summary,insert size and wgs metrics.
- Flagstats are also calculated via samtools for each sample.
- The final outputs are merged metrics from all samples within a `seq_type`. 
- By default, the `_picard_qc_merge_metrics` rule takes `config["lcr-modules"]["_shared"]["samples"]`. To modify, set `CFG["samples]` in the project Snakefile using:
    ```python
    config["lcr-modules"]["picard_qc"]["samples"] = SAMPLES
    ```
- Users are required to specify `CFG['inputs']['intervals']`, which refers to a text file containing the capture regions used in sequencing.
- Reference files for CollectRNaSeq rRNA interval list and refFlat files) are specified in `CFG["inputs"]`. By default it refers to the files generated by `reference_files` subworkflow.
